File: src/main/java/moe/kurenai/demo/HelperAgent.java
package moe.kurenai.demo;

import net.bytebuddy.agent.builder.AgentBuilder;
import net.bytebuddy.asm.Advice;
import net.bytebuddy.implementation.MethodDelegation;
import net.bytebuddy.matcher.ElementMatchers;

import java.lang.instrument.Instrumentation;

public class HelperAgent {

    public static void premain(String args, Instrumentation inst) {
        System.out.println("Starting the agent...");
        new AgentBuilder.Default()
                .type(ElementMatchers.nameContains("ActivationManager"))
                .transform((builder, typeDescription, classLoader, module, domain) ->
                        builder.method(ElementMatchers.any())
                                .intercept(MethodDelegation.to(HelperInterceptor.class)
//                        builder.visit(Advice.to(HelperAdvice.class).on(ElementMatchers.named("isActivated"))
                                )).installOn(inst);
        new AgentBuilder.Default()
                .type(ElementMatchers.nameContains("ActivationManager"))
                .transform((builder, typeDescription, classLoader, module, domain) ->
                        builder.method(ElementMatchers.any())
                                .intercept(MethodDelegation.to(HelperInterceptorVoid.class)
                                )).installOn(inst);
        new AgentBuilder.Default()
                .type(ElementMatchers.nameContains("RSAUtils"))
                .transform((builder, typeDescription, classLoader, module, domain) ->
                        builder.method(ElementMatchers.any())
                                .intercept(MethodDelegation.to(HelperInterceptor.class)
                                )).installOn(inst);
    }

}

File: src/main/kotlin/HelperAgent.kt
import net.bytebuddy.agent.builder.AgentBuilder
import net.bytebuddy.asm.Advice
import net.bytebuddy.description.NamedElement
import net.bytebuddy.description.method.MethodDescription
import net.bytebuddy.description.type.TypeDescription
import net.bytebuddy.dynamic.DynamicType
import net.bytebuddy.matcher.ElementMatchers
import net.bytebuddy.utility.JavaModule
import java.lang.instrument.Instrumentation

object HelperAgent {

    fun premain(agentArgs: String?, inst: Instrumentation) {
        println("Starting the agent...")
        AgentBuilder.Default()
            .type(ElementMatchers.nameContains("ActivationManager"))
            .transform { builder, typeDescription, classLoader, module, protectionDomain ->
                builder.visit(
                    Advice.to(
                        HelperAdvice::class.java
                    ).on(
                        ElementMatchers.named<NamedElement>("isActivated")
                    )
                )
            }.installOn(inst)
    }

}